File: Study/인접행렬/4. 맵과 지도/1인곳 탐색.swift
//
//  main2.swift
//  Algorithm
//
//  Created by 김동현 on 1/15/25.
//

import Foundation

/*
 - 왼쪽 상단부터 시작(0, 0은 무조건 1임을 보장)
 - 4방향 탐색 가능, 방문한 정점은 다시 방문하지 않는다   -> dydx, visited
 - 정점을 다 방문하면서(1인 곳만 갈 수 있음 )
 - 방문한 정점 출력
 
   1 0 1
   1 0 1
   0 1 1
 */

// 상수
let n: Int = 3
let dy: [Int] = [-1, 0, 1, 0]
let dx: [Int] = [0, 1, 0, -1]

// 입력 데이터와 방문 배열
var a: [[Int]] = []
var visited: [[Bool]] = Array(repeating: Array(repeating: false, count: n), count: n)

func go(_ y: Int, _ x: Int) {
    visited[y][x] = true
    print("\(y) : \(x)")
    for i in 0..<4 {
        let ny = y + dy[i]
        let nx = x + dx[i]
        
        if ny<0 || ny>=n || nx<0 || nx>=n { continue }
        if a[ny][nx] == 0 { continue }
        if visited[ny][nx] { continue }
        go(ny, nx)
    }
}

@main
struct Main {
    static func main() {
        freopen("input.txt", "r", stdin);
        for _ in 0..<n {
            let line = readLine()!.split(separator: " ").map { Int($0)! }
            a.append(line)
        }
        go(0, 0)
    }
}

File: 0. 알고리즘 개념공부/2주차/04. 맵과 지도/1인곳 탐색.cpp
//
//  main2.cpp
//  Algorithm
//
//  Created by 김동현 on 1/15/25.
//

/*
 - 왼쪽 상단부터 시작(0, 0은 무조건 1임을 보장)
 - 4방향 탐색 가능, 방문한 정점은 다시 방문하지 않는다   -> dydx, visited
 - 정점을 다 방문하면서(1인 곳만 갈 수 있음 )
 - 방문한 정점 출력
 
   1 0 1
   1 0 1
   0 1 1
 */

#include <iostream>
using namespace std;


const int n = 3;
int a[n][n], visited[n][n];
const int dy[] = {-1, 0, 1, 0};
const int dx[] = {0, 1, 0, -1};

void go(int y, int x) {
    visited[y][x] = 1;
    cout << y << " : " << x << "\n";
    
    // 4방향탐색
    for (int i=0; i<4; i++) {
        int ny = y + dy[i];
        int nx = x + dx[i];
        
        // 오버플로우 / 언더플로우 방지
        // 배열을 마이너스로 참고 못함, 선언한 배열을 넘어서 접근하지 못함
        if (ny<0 || ny>=n || nx<0 || nx>=n) continue;
        
        // 0은 갈 수 없다
        if (a[ny][nx] == 0) continue;
        
        // 방분한 정점은 갈 수 없다
        if (visited[ny][nx]) continue;
        go(ny, nx);
        
    }
}

int main() {
    freopen("input.txt", "r", stdin);
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            cin >> a[i][j];
        }
    }
    go(0, 0);
    return 0;
}

